[WIP] Calculate bezier curve length

danieldouglas92 opened the pull request:
Add the functionality within `bezier_curve.cc` to determine the total arc-length of the bezier curve across all user-specified coordinates, and the length of the bezier curve between each incremental pair of user-specified coordinates.

danieldouglas92 commented:
At this point I think the code that I have is correct, there was an issue with the existing implementation of the Bezier curve when dealing with points that were co-linear.

If you have points P1 and P2, to compute the Bezier curve two additional control points CP1 and CP2 must be determined which control the shape of the curve. Sometimes, the control points would be outside of the bounds of Points P1 and P2, as illustrated below:

x----o--------x----o
P1    CP1        P2     CP2

However, the control points should lie within on the vector P1P2, not beyond it, like this:
x----o--------o----x
P1    CP1        CP2    P2

This is important because when the control points lie beyond the vector P1P2, the arclength of the bezier curve is artifically made longer.

Currently, some of the tests are still failing but they are mostly related to tests which utilize coordinates that are colinear (with the exception of the test `slab_interpolation_simple_curved_CMS.dat`

MFraters reviewed (commented):
This looks good to me. thanks for finding and fixing the bug! I have a few small comments, but otherwise I think it is good to go.

danieldouglas92 commented on include/world_builder/features/subducting_plate.h:
```diff
@@ -236,4 +236,4 @@ namespace WorldBuilder
   } // namespace Features
 } // namespace WorldBuilder
 
-#endif
```
I'm not sure what is happening here? This is the same as the current file and I made no changes to it.

danieldouglas92 commented:
Thanks for the review! I've addressed your comments, it turns out it wasn't `slab_interpolation_simple_curved_CMS` that was failing, but actually `slab_interpolation_simple_CMS` that was failing. This means that all the tests that are actually failing currently involve coordinates that are colinear. If you agree that it is ok to update the test results then I will do that in a follow up commit and hopefully that will wrap this up.

danieldouglas92 commented:
After talking to @MFraters I've tentatively updated the test results. The test results should be more correct now, because the tests that were failing involved colinear points. However, we want to do a little more checking to make sure that the new code still produces correct results, so do not merge this quite yet.

danieldouglas92 commented:
I've fixed the merge conflicts and I'll fix the broken tests sometime tomorrow.

MFraters commented:
Thanks! It does seem to fix #861, but I need to look at bit more at it to remember what this pull request is doing. I will have another look at it this Friday.

MFraters commented:
> For the `dat_slab_interpolation_simple_CMS` test, it's less clear, though I suspect this might just be due to the trench coordinates used in the test. The coordinates jump from values of single meters to values of 1000's of meters:

So, does the issue disappear when you use values on the same order of magnitude and/or change it to:

```
[0,0],[50e3,50e3],[125e3,75e3]
```

MFraters commented on source/world_builder/objects/bezier_curve.cc:
```diff
@@ -560,5 +672,6 @@ namespace WorldBuilder
         }
       return closest_point_on_curve;
     }
+
```
?
